crypto/ocsp/ocsp_test.cc

```diff
@@ -1115,3 +1115,67 @@ TEST_P(OCSPURLTest, OCSPParseURL) {
     EXPECT_FALSE(path);
   }
 }
+
+TEST(OCSPTest, OCSPBIOTest) {
+  std::string reqData =
+      GetTestData(std::string("crypto/ocsp/test/aws/ocsp_request.der").c_str());
+  std::vector<uint8_t> ocsp_request_data(reqData.begin(), reqData.end());
+  std::string respData = GetTestData(
+      std::string("crypto/ocsp/test/aws/ocsp_response.der").c_str());
+  std::vector<uint8_t> ocsp_response_data(respData.begin(), respData.end());
+
+  // Write an OCSP request into the BIO.
+  bssl::UniquePtr<BIO> bio(BIO_new(BIO_s_mem()));
+  int reqLen = (int)ocsp_request_data.size();
+  ASSERT_EQ(BIO_write(bio.get(), ocsp_request_data.data(), reqLen), reqLen);
+  bssl::UniquePtr<OCSP_REQUEST> ocspRequest(
+      d2i_OCSP_REQUEST_bio(bio.get(), nullptr));
+  EXPECT_TRUE(ocspRequest);
+
+  // Reserialize the OCSP request.
+  EXPECT_TRUE(i2d_OCSP_REQUEST_bio(bio.get(), ocspRequest.get()));
+  const uint8_t *bio_data;
+  size_t bio_len;
+  ASSERT_TRUE(BIO_mem_contents(bio.get(), &bio_data, &bio_len));
+  EXPECT_EQ(Bytes(bio_data, bio_len), Bytes(ocsp_request_data.data(), reqLen));
+
+  // Write an OCSP response into the BIO, but it shouldn't successfully parse
+  // with |d2i_OCSP_REQUEST_bio|.
+  BIO_reset(bio.get());
+  const int respLen = (int)ocsp_response_data.size();
+  ASSERT_EQ(BIO_write(bio.get(), ocsp_response_data.data(), respLen), respLen);
+  ocspRequest.reset(d2i_OCSP_REQUEST_bio(bio.get(), nullptr));
+  EXPECT_FALSE(ocspRequest);
+}
+
+TEST(OCSPTest, CertIDDup) {
+  bssl::UniquePtr<X509> issuer(CertFromPEM(
+      GetTestData(std::string("crypto/ocsp/test/aws/ca_cert.pem").c_str())
+          .c_str()));
+  bssl::UniquePtr<X509> subject(CertFromPEM(
+      GetTestData(std::string("crypto/ocsp/test/aws/server_cert.pem").c_str())
+          .c_str()));
+
+  // Create a sample OCSP_CERTID structure
+  bssl::UniquePtr<OCSP_CERTID> orig_id(
+      OCSP_cert_to_id(EVP_sha256(), subject.get(), issuer.get()));
+  ASSERT_TRUE(orig_id);
+
+  // Duplicate the OCSP_CERTID structure
+  bssl::UniquePtr<OCSP_CERTID> dup_id(OCSP_CERTID_dup(orig_id.get()));
+  ASSERT_TRUE(dup_id);
+
+  // Check that the duplicated structure has the same values as the original.
```

Comment:
I realize this is just calling another ASN1 function so it's unlikely this is an issue. But for the sake of completion can you add a check that dup_id does not point to the same ocsp structure as orig_id. In theory dup could just return the same pointer and this test would still pass.

Reply:
Good point, I added an `EXPECT_NE` at the bottom.